core-tests/src/test/java/org/mule/runtime/core/processor/strategy/AbstractProcessingStrategyTestCase.java

```diff
@@ -0,0 +1,323 @@
+
+  @After
+  public void after() {
+    cpuLight.shutdownNow();
+    blocking.shutdownNow();
+    cpuIntensive.shutdownNow();
+    asyncExecutor.shutdownNow();
+  }
+
+  @Test
+  public void singleCpuLight() throws Exception {
+    flow.setMessageProcessors(singletonList(cpuLightProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertSingleCpuLight();
+  }
+
+  protected abstract void assertSingleCpuLight();
+
+  @Test
+  public void singleCpuLightConcurrent() throws Exception {
+    FirstInvocationLatchedProcessor latchedProcessor = new FirstInvocationLatchedProcessor(CPU_LITE);
+
+    flow.setMessageProcessors(singletonList(latchedProcessor));
+    flow.initialise();
+    flow.start();
+
+    asyncExecutor.submit(() -> process(flow, testEvent()));
+
+    latchedProcessor.awaitFirst();
+    process(flow, testEvent());
+    latchedProcessor.releaseFirst();
+
+    assertSingleCpuLightConcurrent();
+  }
+
+  protected abstract void assertSingleCpuLightConcurrent();
+
+  @Test
+  public void multipleCpuLight() throws Exception {
+    flow.setMessageProcessors(asList(cpuLightProcessor, cpuLightProcessor, cpuLightProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertMultipleCpuLight();
+  }
+
+  protected abstract void assertMultipleCpuLight();
+
+  @Test
+  public void blocking() throws Exception {
+    flow.setMessageProcessors(singletonList(blockingProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertSingleBlocking();
+  }
+
+  protected abstract void assertSingleBlocking();
+
+  @Test
+  public void multipleBlocking() throws Exception {
+    flow.setMessageProcessors(asList(blockingProcessor, blockingProcessor, blockingProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertMultipleBlocking();
+  }
+
+  protected abstract void assertMultipleBlocking();
+
+  @Test
+  public void singleCpuIntensive() throws Exception {
+    flow.setMessageProcessors(singletonList(cpuIntensiveProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertSingleCpuIntensive();
+  }
+
+  protected abstract void assertSingleCpuIntensive();
+
+  @Test
+  public void multipleCpuIntensive() throws Exception {
+    flow.setMessageProcessors(asList(cpuIntensiveProcessor, cpuIntensiveProcessor, cpuIntensiveProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertMultipleCpuIntensive();
+  }
+
+  protected abstract void assertMultipleCpuIntensive();
+
+  @Test
+  public void mix() throws Exception {
+    flow.setMessageProcessors(asList(cpuLightProcessor, cpuIntensiveProcessor, blockingProcessor));
+    flow.initialise();
+    flow.start();
+
+    process(flow, testEvent());
+
+    assertMix();
+  }
+
+  protected abstract void assertMix();
```

Comment:
if the assertion will depend on the actual processors, maybe the list of processors hould also be determined by the subclasses.